File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const N = maps.length;
    const M = maps[0].length;
    // 상하좌우
    const moveX = [-1, 1, 0, 0];
    const moveY = [0, 0, -1, 1];
    
    const visited = Array(N).fill().map(e => Array(M).fill(0));
    visited[0][0] = 1;
    const queue = [[0, 0]];
    
    
    while (queue.length > 0) {
        const [X, Y] = queue.shift();
        
        
        if (X === N - 1 && Y === M - 1) {
            return visited[X][Y];
        }
        for (let i = 0; i < moveX.length; i++) {
            const nextX = X + moveX[i];
            const nextY = Y + moveY[i];
            if (nextX >= 0 && nextX < N && nextY >= 0 && nextY < M && maps[nextX][nextY] === 1 && visited[nextX][nextY] === 0) {
                visited[nextX][nextY] = visited[X][Y] + 1;
                queue.push([nextX, nextY]); 
            }
        }
    }
    
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    queue = deque()
    queue.append((0, 0, 1))
    
    visited = [[False] * m for _ in range(n)]
    visited[0][0] = True
    
    while queue:
        current_x, current_y, current_dist = queue.popleft()
        
        if current_x == n - 1 and current_y == m - 1:
            return current_dist
        
        for i in range(4):
            moving_x = current_x + dx[i]
            moving_y = current_y + dy[i]
            
            if moving_x < 0 or moving_x >= n or moving_y < 0 or moving_y >= m:
                continue
                
            if maps[moving_x][moving_y] == 1 and not visited[moving_x][moving_y]:
                visited[moving_x][moving_y] = True
                queue.append((moving_x, moving_y, current_dist + 1))
                
    return -1